### ISSUE TITLE
wallet.completeTx() adds the same input twice!

### ISSUE THREAD

**Issue by pedr0-fr** [bitcoinj/bitcoinj#2063]
See the code below to replicate the issue

```java
import org.apache.commons.codec.binary.Hex;

public class DoubleInputIssue {

    public static void main(String[] args) throws Exception{
        Path walletPath = Path.of(System.getProperty("user.dir"), "data", "wallet.dat");
        File walletFile = new File(walletPath.toString());
        Wallet wallet = Wallet.loadFromFile(walletFile);

        byte[] model_tx = Hex.decodeHex("0100000001df9b45525d3657c483c2444e3b4e68ad9e7d74d5671bca42498cee49c965249a000000006b483045022100ac2054c509a77a4013449ee2ebd095f451a4ce72cf86687363f76d4820d92dd40220486b752486b8a1631d46b6b54e7c14a3e04cd57ea94b3f6847fe57bcf161854601210295f39e3ab984408746401986c25051cbf993b4535684288e832c6ef890626ac5ffffffff01e8030000000000001976a9143226dda88fda8d9dc88f70636ac5b26d3409aecb88ac00000000");

        SendRequest sr = SendRequest.forTx(new Transaction(TestNet3Params.get(), model_tx));
        sr.setFeePerVkb(Coin.ofSat(10_000));
        wallet.completeTx(sr);
        System.out.println(Hex.encodeHexString(sr.tx.bitcoinSerialize()));
    }
}

```

The code above prints the following transaction:

```json
bitcoin-cli -testnet decoderawtransaction "0100000002df9b45525d3657c483c2444e3b4e68ad9e7d74d5671bca42498cee49c965249a000000006b483045022100ac2054c509a77a4013449ee2ebd095f451a4ce72cf86687363f76d4820d92dd40220486b752486b8a1631d46b6b54e7c14a3e04cd57ea94b3f6847fe57bcf161854601210295f39e3ab984408746401986c25051cbf993b4535684288e832c6ef890626ac5ffffffffdf9b45525d3657c483c2444e3b4e68ad9e7d74d5671bca42498cee49c965249a000000006a47304402204cbd98588e8e97b4dcaa6639c7b11a5e56609e23f75aa1cdb95f527cb260a72f02201b397f8e7246facb479224a365b409ecbc93ceabb75f638bf9ac168a3038ef6d01210295f39e3ab984408746401986c25051cbf993b4535684288e832c6ef890626ac5ffffffff02f88c0700000000001976a914545bfcbe31cd54ec715103f12dc828f15261f30888ace8030000000000001976a9143226dda88fda8d9dc88f70636ac5b26d3409aecb88ac00000000"
{
  "txid": "166c74ff15d4b5193324b6475e7962c29084f29c99bdc2b771e1cb5ff3575f1e",
  "hash": "166c74ff15d4b5193324b6475e7962c29084f29c99bdc2b771e1cb5ff3575f1e",
  "version": 1,
  "size": 373,
  "vsize": 373,
  "weight": 1492,
  "locktime": 0,
  "vin": [
    {
      "txid": "9a2465c949ee8c4942ca1b67d5747d9ead684e3b4e44c283c457365d52459bdf",
      "vout": 0,
      "scriptSig": {
        "asm": "3045022100ac2054c509a77a4013449ee2ebd095f451a4ce72cf86687363f76d4820d92dd40220486b752486b8a1631d46b6b54e7c14a3e04cd57ea94b3f6847fe57bcf1618546[ALL] 0295f39e3ab984408746401986c25051cbf993b4535684288e832c6ef890626ac5",
        "hex": "483045022100ac2054c509a77a4013449ee2ebd095f451a4ce72cf86687363f76d4820d92dd40220486b752486b8a1631d46b6b54e7c14a3e04cd57ea94b3f6847fe57bcf161854601210295f39e3ab984408746401986c25051cbf993b4535684288e832c6ef890626ac5"
      },
      "sequence": 4294967295
    },
    {
      "txid": "9a2465c949ee8c4942ca1b67d5747d9ead684e3b4e44c283c457365d52459bdf",
      "vout": 0,
      "scriptSig": {
        "asm": "304402204cbd98588e8e97b4dcaa6639c7b11a5e56609e23f75aa1cdb95f527cb260a72f02201b397f8e7246facb479224a365b409ecbc93ceabb75f638bf9ac168a3038ef6d[ALL] 0295f39e3ab984408746401986c25051cbf993b4535684288e832c6ef890626ac5",
        "hex": "47304402204cbd98588e8e97b4dcaa6639c7b11a5e56609e23f75aa1cdb95f527cb260a72f02201b397f8e7246facb479224a365b409ecbc93ceabb75f638bf9ac168a3038ef6d01210295f39e3ab984408746401986c25051cbf993b4535684288e832c6ef890626ac5"
      },
      "sequence": 4294967295
    }
  ],
  "vout": [
    {
      "value": 0.00494840,
      "n": 0,
      "scriptPubKey": {
        "asm": "OP_DUP OP_HASH160 545bfcbe31cd54ec715103f12dc828f15261f308 OP_EQUALVERIFY OP_CHECKSIG",
        "hex": "76a914545bfcbe31cd54ec715103f12dc828f15261f30888ac",
        "reqSigs": 1,
        "type": "pubkeyhash",
        "addresses": [
          "moD1FZC6o45z7PNfqa7YC7PRxH4podZcuU"
        ]
      }
    },
    {
      "value": 0.00001000,
      "n": 1,
      "scriptPubKey": {
        "asm": "OP_DUP OP_HASH160 3226dda88fda8d9dc88f70636ac5b26d3409aecb OP_EQUALVERIFY OP_CHECKSIG",
        "hex": "76a9143226dda88fda8d9dc88f70636ac5b26d3409aecb88ac",
        "reqSigs": 1,
        "type": "pubkeyhash",
        "addresses": [
          "mk68e6JzRX8ZFY3Jtv2Fv1RijnmuDTJXuK"
        ]
      }
    }
  ]
}

```

The wallet.completeTx() method adds an input that already existed... If you can't replicate the issue yourself I can send you the wallet.dat file from bitcoinj

I should add that this method also fails to detect a change output that may already exist in the SendRequest and therefore will also add a duplicate output!

**Comment by schildbach** [comment]
It would be useful if the above example was written as a unit test. That way, we can use it to reproduce the bug, fix the bug and keep the test as a regression test.

Btw. you can also use `Transaction.toString()`, so the detour via "decoderawtransaction" isn't needed.

**Comment by pedr0-fr** [comment]
My guess is that it doesn't depend on the file.. I was trying to make a fresh wallet and import a UTXO into it but it seems like the wallet if not accepting it?
```
String seedCode = "my word list ...";
        DeterministicSeed seed = new DeterministicSeed(seedCode, null, "", Instant.now().minusMillis(2*24*60*60*1000).getEpochSecond());
        Wallet w2 = Wallet.fromSeed(TestNet3Params.get(), seed);

        String rawTx = "02000000000101a773e33fdd9e946acc45ae6a4f97ff4b5d49ead527625351ec79f7f31bb5c9790100000000fdffffff0220a10700000000001976a9143226dda88fda8d9dc88f70636ac5b26d3409aecb88acca9d0700000000001600145042bf818780eb35f501613daa2f0afa64f0d45902483045022100f31c3365bad47da479a6bbc4debf9aa5110a17d5dc80268d537a15760a312acc02202fe15df16d1b48dec3562beefe3ab7c46f01dfc7f8f487af28ea4cdb7fe03c55012102f50e10c42f500a41a20558ec6fa278d6128935960c7d1e4c088c6954b8bd9ce200000000";
        byte[] serialized = Hex.decodeHex(rawTx);

        Transaction utxo = new Transaction(MainNetParams.get(), serialized);
        WalletTransaction wt = new WalletTransaction(WalletTransaction.Pool.UNSPENT, utxo);
        w2.addWalletTransaction(wt);
        w2.getBalance(); // is zero
       w2.getUnspents(); // is empty
```

I get an InsufficientBalanceException (or whatever is called) before I can check... Checking with a new wallet is going to take some time, because as you can see from my other issue, bitcoinJ takes a long time to find a peer it can download from.

**Comment by pedr0-fr** [comment]
Just tried it without the wallet file and the problem is still there.

**Comment by micheal-swiggs** [comment]
@pedr0-fr what are you trying to do? The deserialised tx sent into `completeTx` currently has the values for its inputs ignored. This seems to be somewhat by design as one may want to send a transaction with inputs signed elsewhere but the value of the transaction should come from inputs contained within the wallet. Obviously the same input should not be added twice. Are you intending for the input value from your transaction to contribute to the output values? Or should that come from other inputs within the wallet?

**Comment by msgilligan** [comment]
Thanks to the work @micheal-swiggs did on PR #3055, I think I understand what is happening:

1. An "incomplete" transaction with an input with an "unconnected" outpoint is passed to `Wallet.completeTx`
2. Since the "unconnected" outpoint doesn't contain value information, `Wallet.completeTx` looks for coins to fund the transaction and re-adds the same input, but this time with "connected" transaction info.

I am working on a simplified version of PR #3055 that should resolve the issue.

I am also wondering if PR #3070 should be added to reject inputs with unconnected outputs.

**Comment by msgilligan** [comment]
PR #3085 is ready-for-review and seems to fix this issue.

**Comment by msgilligan** [comment]
PR #3085 has been merged to the `master` branch.

**Comment by msgilligan** [comment]
I'm going to close this as fixed by #3085. If the problem still occurs with release 0.17-alpha2 or later, please re-open and let us know what is happening.
